horqrux/circuit.py

```diff
@@ -1,47 +1,77 @@
 from __future__ import annotations
 
-from dataclasses import dataclass
+from collections import Counter
+from dataclasses import dataclass, field
 from typing import Any, Callable
 from uuid import uuid4
 
+import jax
+import jax.numpy as jnp
 from jax import Array
 from jax.tree_util import register_pytree_node_class
 
+from horqrux.adjoint import ad_expectation, adjoint_expectation
 from horqrux.apply import apply_gate
 from horqrux.parametric import RX, RY, Parametric
 from horqrux.primitive import NOT, Primitive
-from horqrux.utils import zero_state
+from horqrux.utils import DiffMode, zero_state
 
 
 @register_pytree_node_class
 @dataclass
-class Circuit:
-    """A minimalistic circuit class to store a sequence of gates."""
+class QuantumCircuit:
+    """A minimalistic circuit class to store a sequence of gates.
 
-    n_qubits: int
-    feature_map: list[Primitive]
-    ansatz: list[Primitive]
+    Attributes:
+        n_qubits (int): Number of qubits.
+        operations (list[Primitive]): Operations defining the circuit.
+        feature_map_parameters (list[str]): List of parameters that are considered
+            non trainable, used for passing fixed input data to a quantum circuit.
+                The corresponding operations compose the `feature map`.
+    """
 
-    def __post_init__(self) -> None:
-        self.state = zero_state(self.n_qubits)
+    n_qubits: int
+    operations: list[Primitive]
+    feature_map_parameters: list[str] = field(default_factory=list)
 
-    def __call__(self, param_values: Array) -> Array:
-        return apply_gate(
-            self.state,
-            self.feature_map + self.ansatz,
-            {name: val for name, val in zip(self.param_names, param_values)},
-        )
+    def __call__(self, state: Array, values: dict[str, Array]) -> Array:
+        if state is None:
+            state = zero_state(self.n_qubits)
+        return apply_gate(state, self.operations, values)
 
     @property
     def param_names(self) -> list[str]:
-        return [str(op.param) for op in self.ansatz if isinstance(op, Parametric)]
+        """List of parameters of the circuit.
+            Composed of variational and feature map parameters.
+
+        Returns:
+            list[str]: Names of parameters.
+        """
+        return [str(op.param) for op in self.operations if isinstance(op, Parametric)]
+
+    @property
+    def variational_param_names(self) -> list[str]:
```

Comment:
```suggestion
    def vparam_names(self) -> list[str]:
```

Maybe a similar method for feature parameter names ?

Reply:
Gotcha, will do. Will just rename to `fparams` the input then and rename to `vparams` the method.